technology_and_tooling/testing/testable_code_fixtures.md

```diff
@@ -0,0 +1,799 @@
+    patient = Patient(id=1, data=[1, 2, 3, 4, 5])
+    assert patient.data_min() == 1
+
+def test_patient_attributes():
+    patient = Patient(id=2, data=[10, 20, 30, 40, 50])
+    assert patient.id == 2
+    assert patient.data == [10, 20, 30, 40, 50]
+~~~
+
+:::
+::::
+
+In the exercise above, we found ourselves having to create the same or similar `Patient` objects multiple times. To prevent this repetition, we could encapsulate these tests in their own class, `TestPatient`. Writing tests in this manner helps to organise similar tests into groups and also allows sharing of data between tests. The `pytest` library defines methods that you can add to your class, such as `setup_class` which will be run before running all of the tests in that class or `setup_method` that will be run before each test within the class. This method can be used for creating data or opening files, for example. An additional method called `teardown_class` could be also be added, if needed, and `pytest` will run this method after the tests in the class have completed. Alternatively `teardown_method` will run after each test. These methods can be useful for cleaning up in cases where files were created on your system or an connections were opened. For more information you can [view the documentation here](https://docs.pytest.org/en/latest/how-to/xunit_setup.html).
+
+
+::::challenge{id=test-patient-class title="Grouping tests in a class."}
+
+ Encapsulate the tests for the `Patient` class inside a class named `TestPatient`. Include a method `setup_class` where the two `Patient` objects (with `id` of 1 and 2) will be created rather than creating an object within each test.
+
+:::solution
+~~~python
+import pytest
+from inflammation.models import Patient
+
+class TestPatient:
+    def setup_class(self):
+        self.patient1 = Patient(id=1, data=[1, 2, 3, 4, 5])
+        self.patient2 = Patient(id=2, data=[10, 20, 30, 40, 50])
+
+    def test_patient_data_mean(self):
+        assert self.patient1.data_mean() == 3.0
+
+    def test_patient_data_max(self):
+        assert self.patient1.data_max() == 5
+
+    def test_patient_data_min(self):
+        assert self.patient1.data_min() == 1
+
+    def test_patient_attributes(self):
+        assert self.patient2.id == 2
+        assert self.patient2.data == [10, 20, 30, 40, 50]
+
+~~~
+
+:::
+::::
+
+:::callout{variant="note"}
+## Fixtures
+
+As an alternative to encapsulating test methods in a class and using `setup` and `teardown` methods, we can use *fixtures*. Fixtures are defined by using the `@pytest.fixture` decorator on a function. This function will then become available to be passed as an argument to your tests and used within them. If there is a cleanup part to the code, then the fixture function should use a `yield` statement rather than a `return` statement. Anything up to the `yield` statement is setup code, and anything after the statement will be run post-testing in order to clean up (teardown code).
```

Comment:
I feel like the discussion on `yield` would only confuse students (this is very different control flow than what they are used to), and its not used in the example below so probably can be removed?

Reply:
ahhh no, I see later on you use `yield`. So perhaps move this description closer to where you use it?

Reply:
Have moved this, thanks!